### PR TITLE
contracts-bedrock: add tests for preinstalls

### PR DESCRIPTION
**Pull request by tynes**
**Description**

This is a prereq of https://github.com/ethereum-optimism/optimism/pull/8942.
This will help to remove the concept of `check-l2` to help reduce the
amount of diff a developer needs to do when making changes to the
solidity code. Ideally, a solidity dev needs to touch 0 code in any
other language (besides bash ffi lolz).

After this PR as well as:
- https://github.com/ethereum-optimism/optimism/pull/8933
- https://github.com/ethereum-optimism/optimism/pull/8911

the functionality of `check-l2` is now completely in solidity.

Note that we are now calling the non system contracts "preinstalls"
instead of "predeploys" to denote that they do not live in the predeploy
namespace and are not critical to the functionality of the system.
They are set at deterministic addresses.

### CODE REVIEW

**Comment by tynes** [comment]
The hardcoded bytecode was pulled from the bytecode in the `_more.go` files which are used to populate the genesis

**Comment by refcell** [comment]
> The hardcoded bytecode was pulled from the bytecode in the `_more.go` files which are used to populate the genesis

Is it possibly to read in the bytecode from the `_more.go` files dynamically so we don't have hardcoded hex strings?
Do we want to be explicitly asserting the code?

**Comment by tynes** [comment]
> > The hardcoded bytecode was pulled from the bytecode in the `_more.go` files which are used to populate the genesis
> 
> Is it possibly to read in the bytecode from the `_more.go` files dynamically so we don't have hardcoded hex strings? Do we want to be explicitly asserting the code?

Good question - the bytecode should always be the same and we are planning on moving away from using the `_more.go` files to generate the genesis, so I decided to move away from them as a source of truth. The `_more.go` files have the bytecode hardcoded. A future PR will migrate the hardcoded values to a library and that library will be used for genesis generation. Will not delete the hardcoded values in the tests until the see it working the new way. permit2 differs per chain so we check that its chainid was correctly set

File: packages/contracts-bedrock/test/Predeploys.t.sol [hunk 1]
```diff
@@ -71,4 +75,86 @@ contract PredeploysTest is CommonTest {
...
+    function test_preinstall_permit2_succeeds() external {
+        bytes32 domainSeparator = IEIP712(Predeploys.Permit2).DOMAIN_SEPARATOR();
+        bytes32 typeHash =
+            keccak256(abi.encodePacked("EIP712Domain(string name,uint256 chainId,address verifyingContract)"));
+        bytes32 nameHash = keccak256(abi.encodePacked("Permit2"));
+        uint256 chainId = block.chainid;
+        bytes memory encoded = abi.encode(typeHash, nameHash, chainId, Predeploys.Permit2);
+        bytes32 expectedDomainSeparator = keccak256(encoded);
+        assertEq(domainSeparator, expectedDomainSeparator, "Domain separator mismatch");
+    }
```
Comment (refcell) [inline packages/contracts-bedrock/test/Predeploys.t.sol]: nit: Would you mind moving this test above code assertions (`test_preinstall_multicall3_succeeds` test looks to be the first one) and put a divider so it's clear where code assertion tests should go vs other tests? Just makes it easier to find this and other tests given large hex strings

ie
```
    function test_preinstall_permit2_succeeds() external { ... }

    //////////////////////////////////////////////////////
    /// Code Assertion Tests
    //////////////////////////////////////////////////////

    function test_preinstall_multicall3_succeeds() external { ... }
```

**Review by refcell — CHANGES_REQUESTED** [review]
1 small nit

File: packages/contracts-bedrock/test/Predeploys.t.sol (hunk 1 above)
Comment (tynes) [reply packages/contracts-bedrock/test/Predeploys.t.sol]: i have addressed the comment :D

**Review by refcell — APPROVED** [review]
LGTM

**Pull request merged**